core/src/apps/management/sd_protect.py

```diff
@@ -13,23 +14,33 @@
     request_pin_and_sd_salt,
     show_pin_invalid,
 )
-from apps.common.sd_salt import (
-    SD_SALT_AUTH_KEY_LEN_BYTES,
-    SD_SALT_AUTH_TAG_LEN_BYTES,
-    SD_SALT_LEN_BYTES,
-    commit_sd_salt,
-    remove_sd_salt,
-    set_sd_salt,
-    stage_sd_salt,
-)
-from apps.common.storage import device, is_initialized
+from apps.common.sd_salt import ensure_sd_card, sd_write_failed_dialog
 
 if False:
+    from typing import Awaitable, Tuple
     from trezor.messages.SdProtect import SdProtect
 
 
+def _make_salt() -> Tuple[bytes, bytes, bytes]:
+    salt = random.bytes(storage.sd_salt.SD_SALT_LEN_BYTES)
+    auth_key = random.bytes(storage.device.SD_SALT_AUTH_KEY_LEN_BYTES)
+    tag = storage.sd_salt.compute_auth_tag(salt, auth_key)
+    return salt, auth_key, tag
+
+
+async def _set_salt(
+    ctx: wire.Context, salt: bytes, salt_tag: bytes, stage: bool = False
+) -> None:
+    while True:
```

Comment:
Well, actually not. I hate the fact that `sd.power(False)` is called in a totally different place than `sd.power(True)`. This makes it very difficult to keep track so that it doesn't get skipped. I think that `ensure_sd_card()` should be inside storage/sd_salt.py, but I understand, that that's a problem, so I propose this:
- Get rid of `ensure_sd_card()`.
- In storage/sd_salt.py `if not sd.power(True):` will raise a new exception like `SdCardMissing` instead of `OSError`.
- The `while` loop here (and in `request_sd_salt`) will handle `SdCardMissing` by displaying the appropriate dialog.

Reply:
i'm not sure if this is better than moving `ensure_sd_card` here?

I actually expect `ensure_sd_card` to get moved to something like `apps.common.sd_card` as soon as we have more SD-card-based features, so that it is usable generally, without the need to handle some `SdCardMissing` code explicitly.

The original reasoning for `ensure_sd_card` in the calling function, instead of `_set_sd_card`, is that afterwards you call, e.g., `storage.sd_salt.commit()`. My logic is that after `ensure_sd_salt`, you can safely call SD-related functions.

OTOH maybe functions in `storage.sd_salt` should not be touching SD power at all? And the caller should be responsible for that, and `fs.mount()` will presumably fail with an IOError or something if SD card is not enabled.

Reply:
I thought the whole point of this major refactor was to separate the IO logic from the app logic. So how does it make sense to handle powering the SD card on (and possibly off) in the app logic? That seems way worse than keeping the two coupled.

There are two reasons why I want to keep the `sd.power()` calls localized and restricted as much as possible:
1. We need to make sure that the SD card is not left powered on, which for example would have happened if the user cancelled PIN entry in https://github.com/trezor/trezor-firmware/pull/665/files#diff-1e29cd2c01a1174872a782f80bb6d78aR138. This is already very difficult to review even when you know what you are looking for.
2. I was lead to believe that when the SD card is powered on it drains non-negligible power, so I wanted to keep that to a minimum for when we might use this on a device with a battery. For example, it makes no sense to keep the SD card powered-on during error dialogs or during PIN entry. @prusnak do you have a better idea of how significant the power drainage is?

Reply:
my question here is more like: is it OK to turn the card on and off in each individual call to `storage.sd_salt.*`?

Of course, we would need to encapsulate the SD power properly. Like a context manager: `with sd_powered_on(): ...`

Reply:
> my question here is more like: is it OK to turn the card on and off in each individual call to `storage.sd_salt.*`?

Good question, I am actually not sure. @stick?

> Of course, we would need to encapsulate the SD power properly. Like a context manager: `with sd_powered_on(): ...`

Good idea, solves point 1, but doesn't solve point 2.

Reply:
One more question for @prusnak, is it a problem if the SD card is removed while io.SDCard is powered on?

Reply:
It's certainly not nice and might cause reboot with TT. This is fixed in the later edition of the board. However, the code should recover correctly in this case. If it does not then it is a bug and we should fix it.

Reply:
how about turning the power on-and-off frequently?